core/lib/zksync_core/src/metadata_calculator/helpers.rs

```diff
@@ -355,15 +354,15 @@ impl LazyAsyncTreeReader {
     }
 
     /// Waits until the tree is initialized and returns a reader for it.
-    pub async fn wait(mut self) -> AsyncTreeReader {
+    pub async fn wait(mut self) -> anyhow::Result<AsyncTreeReader> {
```

Comment:
Maybe an `Option` would be better? What I don't quite like in the current approach is that it'll produce errors during shutdown whereas logically they are not errors (the corresponding task should early-return `Ok(())` if it encounters such an "error"). OTOH, this is significantly less fluent than using `?`. One way to improve it (albeit fragile) is to create a special error type for such early returns and detect it in [the error chain](https://docs.rs/anyhow/latest/anyhow/struct.Error.html#method.chain) of task errors (if it's present, the task hasn't *really* errored).

Reply:
I thought about option too, but then I thought that maybe it _is_ an error to wait for a tree if there never will be one? This way we'll get an error report about potentially useless operation -- which may be a good thing.